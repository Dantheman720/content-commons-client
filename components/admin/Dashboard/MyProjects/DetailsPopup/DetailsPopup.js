import React, { useState } from 'react';
import PropTypes from 'prop-types';
import dynamic from 'next/dynamic';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import debounce from 'lodash/debounce';
import { Popup } from 'semantic-ui-react';
import ApolloError from 'components/errors/ApolloError';
import './DetailsPopup.scss';

const CHECK_PROJECT_TYPE_QUERY = gql`
  query CheckProjectType( $id: ID! ) {
    videoProject( id: $id ) {
      id
      projectType
    }
  }
`;

const VideoDetailsPopup = dynamic( () => import( './VideoDetailsPopup' ) );
const ImageDetailsPopup = dynamic( () => import( './ImageDetailsPopup' ) );

const renderPopup = ( projectType, id ) => {
  if ( projectType === 'VideoProject' ) {
    return <VideoDetailsPopup id={ id } />;
  }
  if ( projectType === 'ImageProject' ) {
    return <ImageDetailsPopup id={ id } />;
  }
};

const DetailsPopup = props => {
  const { id } = props;
  const [detailsPopupOpen, setDetailsPopupOpen] = useState( false );

  const handleResize = debounce( () => {
    /* eslint-disable no-use-before-define */
    handleClose();
  }, 500, { leading: true, trailing: false } );

  const handleTableScroll = debounce( () => {
    /* eslint-disable no-use-before-define */
    handleClose();
  }, 500, { leading: true, trailing: false } );

  const handleOpen = () => {
    setDetailsPopupOpen( true );
    window.addEventListener( 'resize', handleResize );

    const itemsTable = document.querySelector( '.items_table' );
    itemsTable.addEventListener( 'scroll', handleTableScroll );
  };

  const handleClose = () => {
    setDetailsPopupOpen( false );
    window.removeEventListener( 'resize', handleResize );

    const itemsTable = document.querySelector( '.items_table' );
    itemsTable.removeEventListener( 'scroll', handleTableScroll );
  };

  return (
    <Query query={ CHECK_PROJECT_TYPE_QUERY } variables={ { id: props.id } }>
      {
        ( { loading, error, data } ) => {
          if ( loading ) return <p>Loading....</p>;
          if ( error ) return <ApolloError error={ error } />;

          if ( data.videoProject ) {
            const { __typename } = data.videoProject;
            return (
              // 06/10/19 - Updating button text from "Details" to "Files"
              // if DetailsPopup will contain content other than files in future,
              // will add conditional to display "Details" text along with "Files"/"Other Content"
              // subheaders within popup
              <Popup
                className="detailsFiles_popup"
                trigger={ (
                  <button
                    type="button"
                    className="linkStyle myProjects_data_actions_action"
                    data-projectitempopup="detailsPopup"
                  >
                    Files
                  </button>
                ) }
                content={ renderPopup( __typename, id ) }
                on="click"
                position="bottom left"
                keepInViewPort
                open={ detailsPopupOpen }
                onOpen={ handleOpen }
                onClose={ handleClose }
              />
            );
          }

          return <p>There are no supporting files for this project.</p>;
        }
      }
    </Query>
  );
};

DetailsPopup.propTypes = {
  id: PropTypes.string
};

export default DetailsPopup;
