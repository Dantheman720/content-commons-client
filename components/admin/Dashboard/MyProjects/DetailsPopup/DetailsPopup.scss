@import "../../../../../styles/colors.scss";

.ui.popup.detailsFiles_popup {
  max-width: inherit;
  padding: 0;
  color: $shuttle-grey;
  background-color: $light-grey;
  @media screen and (max-width: 400px) {
    margin-left: 1em;
  }

  &:before {
    z-index: -1;
    background-color: $light-grey !important;
  }

  p {
    margin-bottom: 3px;
    font-size: 11px;  
  }

  .details-files {
    display: flex;
    flex-wrap: wrap;
    max-width: 300px;
    max-height: 200px;
    overflow-y: auto;
    padding: 0;
  }

  ul {
    padding-left: 0;
    margin: 0;
    font-size: 11px;
    line-height: 1.6;
    list-style: none;

    li {
      padding: 1em;
      
      &:not(:last-child) {
        border-bottom: 1px solid hsla( 0, 0, 50%, 0.2 );
      }      
    }
  }
}


