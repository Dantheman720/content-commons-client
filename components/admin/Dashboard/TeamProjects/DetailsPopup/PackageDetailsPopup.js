import React from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import ApolloError from 'components/errors/ApolloError';
import { formatBytes, getCount } from 'lib/utils';

import { packageDocsMock } from './packageDocsMock';

const PACKAGE_DOCUMENTS_QUERY = gql`
  query PackageDocumentsQuery( $id: ID! ) {
    package( id: $id ) {
      id
      documents {
        id
        filename      
      }
    }
  }
`;

// const PackageDetailsPopup = props => (
//   <Query query={ PACKAGE_DOCUMENTS_QUERY } variables={ { id: props.id } }>
//     {
//       ( { loading, error, data } ) => {
//         if ( loading ) return <p>Loading....</p>;
//         if ( error ) return <ApolloError error={ error } />;
//         if ( !data.package ) return null;

//         const { documents } = data.package;

//         return (
//           <ul>
//             { documents.map( doc => <li>{ doc.filename }</li> ) }
//           </ul>
//         );
//       }
//     }
//   </Query>
// );

const PackageDetailsPopup = props => {
  const documents = packageDocsMock.result.data;
  return (
    <div className="details-files">
      <ul>
        { documents.map( doc => (
          <li key={ doc.id }>{ doc.filename }</li>
        ) ) }
      </ul>
    </div>
  );
};

PackageDetailsPopup.propTypes = {

};

export default PackageDetailsPopup;
