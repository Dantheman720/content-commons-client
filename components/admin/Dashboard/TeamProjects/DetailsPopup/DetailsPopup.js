import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import dynamic from 'next/dynamic';
import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import debounce from 'lodash/debounce';
import { Popup } from 'semantic-ui-react';
import ApolloError from 'components/errors/ApolloError';
import './DetailsPopup.scss';

const VideoDetailsPopup = dynamic( () => import( './VideoDetailsPopup' ) );
const PackageDetailsPopup = dynamic( () => import( './PackageDetailsPopup' ) );

const CHECK_PROJECT_TYPE_QUERY = gql`
  query CheckProjectType( $id: ID! ) {
    videoProject( id: $id ) {
      id
      projectType
    }
  }
`;

// const CHECK_TYPE_QUERY = gql`
//   query CheckTypeQuery( $id: ID! ) {
//     videoProject( id: $id ) {
//       id
//       projectType
//     }
//     package( id: $id ) {
//       id
//     }
//   }
// `;

const renderPopup = ( projectType, id ) => {
  if ( projectType === 'VideoProject' ) {
    return <VideoDetailsPopup id={ id } />;
  }
  if ( projectType === 'Package' ) {
    return <PackageDetailsPopup id={ id } />;
  }
};

const DetailsPopup = props => {
  const { id } = props;
  const [detailsPopupOpen, setDetailsPopupOpen] = useState( false );

  /**
   * Handle popup behavior on scroll instead of native Popup component hideOnScroll prop
   * Allows to scroll popup (Files) w/ scrollbars w/o immediately closing
   */
  useEffect( () => {
    const handlePopupScroll = () => setDetailsPopupOpen( false );
    window.addEventListener( 'scroll', handlePopupScroll );
    return () => window.removeEventListener( 'scroll', handlePopupScroll );
  }, [] );

  /**
   * Handle memory leak with set state fn call
   * (i.e., setDetailsPopupOpen) on unmounted component.
   */
  let isMounted = true;
  useEffect( () => () => { isMounted = false; }, [] );

  const handleResize = debounce( () => {
    /* eslint-disable no-use-before-define */
    handleClose();
  }, 500, { leading: true, trailing: false } );

  const handleOpen = () => {
    setDetailsPopupOpen( true );
    window.addEventListener( 'resize', handleResize );
  };

  const handleClose = () => {
    window.removeEventListener( 'resize', handleResize );
    const itemsTable = document.querySelector( '.items_table' );
    if ( itemsTable && isMounted ) {
      setDetailsPopupOpen( false );
    }
  };

  const { loading, error, data } = useQuery( CHECK_PROJECT_TYPE_QUERY, {
    variables: { id: props.id }
  } );

  if ( loading ) return <p>Loading....</p>;
  if ( error ) return <ApolloError error={ error } />;
  // if ( !data.videoProject && !data.package ) return null;

  // TEMP
  if ( id === 'package123' ) {
    data.package = {
      __typename: 'Package',
      id,
    };
  }

  // const { __typename } = data.videoProject;
  let __typename;
  if ( data.videoProject != null ) __typename = data.videoProject.__typename;
  if ( data.package != null ) __typename = data.package.__typename;

  return (
    // 06/10/19 - Updating button text from "Details" to "Files"
    // if DetailsPopup will contain content other than files in future,
    // will add conditional to display "Details" text along with "Files"/"Other Content"
    // subheaders within popup
    <Popup
      className="detailsFiles_popup"
      trigger={ (
        <button
          type="button"
          className="linkStyle projects_data_actions_action"
          data-projectitempopup="detailsPopup"
        >
          Files
        </button>
      ) }
      content={ renderPopup( __typename, id ) }
      on="click"
      position="bottom left"
      open={ detailsPopupOpen }
      onOpen={ handleOpen }
      onClose={ handleClose }
    />
  );
};

DetailsPopup.propTypes = {
  id: PropTypes.string
};

export default DetailsPopup;
