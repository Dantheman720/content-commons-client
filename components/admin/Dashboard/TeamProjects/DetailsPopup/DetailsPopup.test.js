import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import DetailsPopup, { CHECK_PROJECT_TYPE_QUERY } from './DetailsPopup';

/**
 * Use custom element for VideoDetailsPopup to avoid
 * "incorrect casing" error triggered by "opens and
 * closes the Popup when the trigger is clicked" test
 * @see https://jestjs.io/docs/en/tutorial-react.html#snapshot-testing-with-mocks-enzyme-and-react-16
 */
// jest.mock( 'next-server/dynamic', () => () => 'VideoDetailsPopup' );
jest.mock( 'next-server/dynamic', () => () => 'video-details-popup' );

const props = { id: '123' };

const mocks = [
  {
    request: {
      query: CHECK_PROJECT_TYPE_QUERY,
      variables: { ...props }
    },
    result: {
      data: {
        videoProject: {
          __typename: 'VideoProject',
          id: props.id,
          projectType: 'LANGUAGE'
        }
      }
    }
  }
];

const Component = (
  <MockedProvider mocks={ mocks } addTypename>
    <DetailsPopup { ...props } />
  </MockedProvider>
);

describe( '<DetailsPopup />', () => {
  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const detailsPopup = wrapper.find( 'DetailsPopup' );
    const loading = <p>Loading....</p>;

    expect( detailsPopup.exists() ).toEqual( true );
    expect( detailsPopup.contains( loading ) ).toEqual( true );
  } );

  it( 'renders error message if an error is returned', async () => {
    const errorMocks = [
      {
        request: {
          query: CHECK_PROJECT_TYPE_QUERY,
          variables: { ...props }
        },
        result: {
          errors: [{ message: 'There was an error.' }]
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename>
        <DetailsPopup { ...props } />
      </MockedProvider>
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const detailsPopup = wrapper.find( 'DetailsPopup' );
    const errorComponent = detailsPopup.find( 'ApolloError' );

    expect( errorComponent.exists() ).toEqual( true );
    expect( errorComponent.contains( 'There was an error.' ) )
      .toEqual( true );
  } );

  it( 'renders the final state', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const detailsPopup = wrapper.find( 'DetailsPopup' );

    expect( toJSON( detailsPopup ) ).toMatchSnapshot();
  } );

  it( 'adds the event listeners on mount', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const map = {};
    window.addEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    setTimeout( () => {
      expect( window.addEventListener ).toHaveBeenCalledWith( 'resize' );
    }, 500 );
  } );

  it( 'removes the event listeners on unmount', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const map = {};
    window.removeEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    setTimeout( () => {
      wrapper.unmount();
      expect( window.removeEventListener ).toHaveBeenCalledWith( 'resize' );
    }, 500 );
  } );

  it( 'renders <VideoDetailsPopup /> as the Popup content', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const detailsPopup = wrapper.find( 'DetailsPopup' );
    const content = detailsPopup.find( 'Popup' ).prop( 'content' );

    expect( content.type ).toEqual( 'video-details-popup' );
    expect( content.props.id ).toEqual( props.id );
  } );

  it( 'renders null if videoProject is falsy', async () => {
    const nullMocks = [
      {
        request: {
          query: CHECK_PROJECT_TYPE_QUERY,
          variables: { ...props }
        },
        result: {
          data: { videoProject: null }
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={ nullMocks } addTypename>
        <DetailsPopup { ...props } />
      </MockedProvider>
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const detailsPopup = wrapper.find( 'DetailsPopup' );

    expect( detailsPopup.html() ).toEqual( null );
  } );

  it( 'clicking the button trigger opens the Popup', async () => {
    const div = document.createElement( 'div' );
    div.classList.add( 'items_table' );
    window.domNode = div;
    document.body.appendChild( div );

    const wrapper = mount( Component, { attachTo: window.domNode } );
    await wait( 0 );
    wrapper.update();

    const popup = () => wrapper.find( 'Popup' );
    const btn = wrapper.find( 'button.projects_data_actions_action' );

    const map = {};
    window.addEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    // popup is initially closed
    expect( popup().prop( 'open' ) ).toEqual( false );

    // open the popup
    btn.simulate( 'click' );
    expect( popup().prop( 'open' ) ).toEqual( true );
  } );

  it( 'global resize event closes the Popup', async () => {
    const div = document.createElement( 'div' );
    div.classList.add( 'items_table' );
    window.domNode = div;
    document.body.appendChild( div );

    const wrapper = mount( Component, { attachTo: window.domNode } );
    await wait( 0 );
    wrapper.update();

    const popup = () => wrapper.find( 'Popup' );
    const btn = wrapper.find( 'button.projects_data_actions_action' );

    const map = {};
    window.addEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    // popup is initially closed
    expect( popup().prop( 'open' ) ).toEqual( false );

    // open the popup
    btn.simulate( 'click' );
    expect( popup().prop( 'open' ) ).toEqual( true );

    /**
     * close the popup
     * Wrap in `setTimeout` as a workaround for `act` warning
     * bug related to `async` action and `useEffect` hook
     * @see https://github.com/facebook/react/issues/14769#issuecomment-507198432
     */
    setTimeout( () => {
      map.resize();
      expect( popup().prop( 'open' ) ).toEqual( false );
    }, 500 );
  } );

  it( 'global click event closes the Popup', async () => {
    const div = document.createElement( 'div' );
    div.classList.add( 'items_table' );
    window.domNode = div;
    document.body.appendChild( div );

    const wrapper = mount( Component, { attachTo: window.domNode } );
    await wait( 0 );
    wrapper.update();

    const popup = () => wrapper.find( 'Popup' );
    const btn = wrapper.find( 'button.projects_data_actions_action' );

    const map = {};
    window.addEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    // popup is initially closed
    expect( popup().prop( 'open' ) ).toEqual( false );

    // open the popup
    btn.simulate( 'click' );
    expect( popup().prop( 'open' ) ).toEqual( true );

    // close the popup
    setTimeout( () => {
      map.click();
      expect( popup().prop( 'open' ) ).toEqual( false );
    }, 500 );
  } );
} );
