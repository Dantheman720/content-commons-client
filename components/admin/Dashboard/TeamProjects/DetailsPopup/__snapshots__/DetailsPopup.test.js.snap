// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailsPopup /> renders the final state 1`] = `
<DetailsPopup
  id="123"
>
  <Query
    query={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "CheckProjectType",
            },
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "videoProject",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "projectType",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "ID",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 107,
          "start": 0,
        },
      }
    }
    variables={
      Object {
        "id": "123",
      }
    }
  >
    <Popup
      className="detailsFiles_popup"
      content={
        <video-details-popup
          id="123"
        />
      }
      keepInViewPort={true}
      on="click"
      onClose={[Function]}
      onOpen={[Function]}
      open={false}
      position="bottom left"
      trigger={
        <button
          className="linkStyle projects_data_actions_action"
          data-projectitempopup="detailsPopup"
          type="button"
        >
          Files
        </button>
      }
    >
      <Portal
        closeOnDocumentClick={true}
        closeOnEscape={true}
        closeOnTriggerClick={true}
        eventPool="default"
        onClose={[Function]}
        onMount={[Function]}
        onOpen={[Function]}
        onUnmount={[Function]}
        open={false}
        openOnTriggerClick={true}
        trigger={
          <button
            className="linkStyle projects_data_actions_action"
            data-projectitempopup="detailsPopup"
            type="button"
          >
            Files
          </button>
        }
        triggerRef={[Function]}
      >
        <Ref
          innerRef={[Function]}
        >
          <button
            className="linkStyle projects_data_actions_action"
            data-projectitempopup="detailsPopup"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="button"
          >
            Files
          </button>
        </Ref>
      </Portal>
    </Popup>
  </Query>
</DetailsPopup>
`;
